src/Compilers/CSharp/Portable/Compilation/MemberSemanticModel.cs

```diff
@@ -1833,10 +1836,41 @@ internal ImmutableArray<BoundNode> GetBoundNodes(CSharpSyntaxNode node)
             var statementBinder = GetEnclosingBinder(GetAdjustedNodePosition(nodeToBind));
             Binder incrementalBinder = new IncrementalBinder(this, statementBinder);
 
+            // PROTOTYPE(nullable-api): Speculative models are going to have to do something more advanced
+            // here. They will need to run nullable analysis up to the point that is being speculated on, and
+            // then take that state and run analysis on the statement or expression being speculated on.
+            // Currently, it will return incorrect info because it's just running analysis on the speculated
+            // part.
+
+            BoundNode root;
             using (_nodeMapLock.DisposableWrite())
             {
-                BoundNode boundStatement = this.Bind(incrementalBinder, nodeToBind, _ignoredDiagnostics);
-                results = GuardedAddBoundTreeAndGetBoundNodeFromMap(node, boundStatement);
+                root = this.Bind(incrementalBinder, nodeToBind, _ignoredDiagnostics);
+            }
+
+            if (EnableNullableAnalysis)
+            {
+                // PROTOTYPE(nullable-api): Handle other symbol types
+                if (MemberSymbol is MethodSymbol method)
+                {
+                    root = NullableWalker.AnalyzeAndRewrite(Compilation, method, root, RootBinder.Conversions, _ignoredDiagnostics);
+                }
+            }
+#if DEBUG
+            else
+            {
+                // Always rewrite in DEBUG mode and just throw the results away.
+                if (MemberSymbol is MethodSymbol method)
+                {
+                    _ = NullableWalker.AnalyzeAndRewrite(Compilation, method, Bind(incrementalBinder, Root, _ignoredDiagnostics), RootBinder.Conversions, _ignoredDiagnostics);
+                }
+            }
+#endif
+
+            using (_nodeMapLock.DisposableWrite())
+            {
+
+                results = GuardedAddBoundTreeAndGetBoundNodeFromMap(node, root);
```

Comment:
>results = GuardedAddBoundTreeAndGetBoundNodeFromMap(node, root); [](start = 16, length = 64)

I am not sure if it is safe to interrupt the lock between binding and adding the result into the map. What if another thread gets in between?

Reply:
Then we would accidentally do the analysis twice, but I don't think we should add to the map twice. `GuardedAddBoundTreeAndGetBoundNodeFromMap` will only add new nodes to the map if they're not already in there, so the thread that gets interrupted would just do some extra work and then throw the results away.

---
In reply to: [269814702](https://github.com/dotnet/roslyn/pull/34429#discussion_r269814702) [](ancestors = 269814702)

Reply:
I'll change this to avoid change existing semantics, but it does feel like this wouldn't be unsafe, just result in more work, which can already happen if two threads happen to get past the initial check for existing results at the same time.

---
In reply to: [271072016](https://github.com/dotnet/roslyn/pull/34429#discussion_r271072016) [](ancestors = 271072016,269814702)